File: slidebacklib/src/main/java/com/parfoismeng/slidebacklib/widget/SlideBackIconView.java
package com.parfoismeng.slidebacklib.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

/**
 * author : ParfoisMeng
 * time   : 2018/12/19
 * desc   : 边缘返回的图标View
 */
public class SlideBackIconView extends View {
    private Path bgPath, arrowPath;
    private Paint bgPaint, arrowPaint;

    private float slideLength = 0; // 拉动距离
    private float maxSlideLength = 0; // 最大拉动距离

    private float arrowSize = 10; // 箭头图标大小

    @ColorInt
    private int backViewColor = Color.BLACK; // 默认值

    private float backViewHeight = 0;

    public SlideBackIconView(Context context) {
        this(context, null);
    }

    public SlideBackIconView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SlideBackIconView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    /**
     * 初始化
     * Path & Paint
     */
    private void init() {
        bgPath = new Path();
        arrowPath = new Path();

        bgPaint = new Paint();
        bgPaint.setAntiAlias(true);
        bgPaint.setStyle(Paint.Style.FILL_AND_STROKE);
        bgPaint.setColor(backViewColor);
        bgPaint.setStrokeWidth(1);

        arrowPaint = new Paint();
        arrowPaint.setAntiAlias(true);
        arrowPaint.setStyle(Paint.Style.STROKE);
        arrowPaint.setColor(Color.WHITE);
        arrowPaint.setStrokeWidth(8);
        arrowPaint.setStrokeJoin(Paint.Join.ROUND);

        setAlpha(0);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // 背景
        bgPath.reset();
        bgPath.moveTo(0, 0);
        bgPath.cubicTo(0, backViewHeight * 2 / 9, slideLength, backViewHeight / 3, slideLength, backViewHeight / 2);
        bgPath.cubicTo(slideLength, backViewHeight * 2 / 3, 0, backViewHeight * 7 / 9, 0, backViewHeight);
        canvas.drawPath(bgPath, bgPaint);

        // 箭头
        float arrowZoom = slideLength / maxSlideLength; // 箭头大小变化率
        float arrowAngle = arrowZoom < 0.75f ? 0 : (arrowZoom - 0.75f) * 2; // 箭头角度变化率
        arrowPath.reset();
        arrowPath.moveTo(slideLength / 2 + (arrowSize * arrowAngle), backViewHeight / 2 - (arrowZoom * arrowSize));
        arrowPath.lineTo(slideLength / 2 - (arrowSize * arrowAngle), backViewHeight / 2);
        arrowPath.lineTo(slideLength / 2 + (arrowSize * arrowAngle), backViewHeight / 2 + (arrowZoom * arrowSize));
        canvas.drawPath(arrowPath, arrowPaint);

        setAlpha(slideLength / maxSlideLength - 0.2f); // 最多0.8透明度
    }

    public void updateSlideLength(float slideLength) {
        this.slideLength = slideLength;
        invalidate();
    }

    /**
     * 设置最大拉动距离
     *
     * @param maxSlideLength px值
     */
    public void setMaxSlideLength(float maxSlideLength) {
        this.maxSlideLength = maxSlideLength;
    }

    /**
     * 设置箭头图标大小
     *
     * @param arrowSize px值
     */
    public void setArrowSize(float arrowSize) {
        this.arrowSize = arrowSize;
    }

    /**
     * 设置返回Icon背景色
     *
     * @param backViewColor ColorInt
     */
    public void setBackViewColor(@ColorInt int backViewColor) {
        this.backViewColor = backViewColor;
    }

    /**
     * 设置返回Icon的高度
     *
     * @param backViewHeight px值
     */
    public void setBackViewHeight(float backViewHeight) {
        this.backViewHeight = backViewHeight;
    }

    public float getBackViewHeight() {
        return backViewHeight;
    }
}

File: slidebacklib/src/main/java/com/parfoismeng/slidebacklib/widget/SlideBackIconView.kt
package com.parfoismeng.slidebacklib.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import kotlin.math.max

/**
 * author : ParfoisMeng
 * time   : 2018/12/19
 * desc   : 边缘返回的图标View
 */
class SlideBackIconView constructor(context: Context?, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {
    // 路径对象
    private val bgPath: Path by lazy { Path() }
    private val arrowPath: Path by lazy { Path() }

    // 画笔对象
    private val bgPaint: Paint by lazy {
        Paint().apply {
            isAntiAlias = true
            style = Paint.Style.FILL_AND_STROKE // 填充内部和描边
            color = viewBackgroundColor
            strokeWidth = 1f // 画笔宽度
        }
    }
    private val arrowPaint: Paint by lazy {
        Paint().apply {
            isAntiAlias = true
            style = Paint.Style.STROKE // 描边
            color = Color.WHITE
            strokeWidth = 8f // 画笔宽度
            strokeJoin = Paint.Join.ROUND // * 结合处的样子 ROUND:圆弧
        }
    }

    var viewBackgroundColor = Color.BLACK // 控件背景色
    var viewHeight = 0f // 控件高度
    var viewArrowSize = 10f // 箭头图标大小
    var viewMaxLength = 0f // 最大拉动距离

    private var currentSlideLength = 0f // 当前拉动距离

    init {
        alpha = 0f
    }

    /**
     * 因为过程中会多次绘制，所以要先重置路径再绘制。
     * 贝塞尔曲线没什么好说的，相关文章有很多。此曲线经测试比较类似“即刻App”。
     *
     *
     * 方便阅读再写一遍，此段代码中的变量定义：
     * iconViewHeight       控件高度
     * currentSlideLength   当前拉动距离
     * iconViewMaxLength    最大拉动距离
     * iconViewArrowSize    箭头图标大小
     */
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // 背景
        bgPath.reset() // 会多次绘制，所以先重置
        bgPath.moveTo(0f, 0f)
        bgPath.cubicTo(0f, viewHeight * 2 / 9, currentSlideLength, viewHeight / 3, currentSlideLength, viewHeight / 2)
        bgPath.cubicTo(currentSlideLength, viewHeight * 2 / 3, 0f, viewHeight * 7 / 9, 0f, viewHeight)
        canvas.drawPath(bgPath, bgPaint) // 根据设置的贝塞尔曲线路径用画笔绘制

        // 箭头是先直线由短变长再折弯变成箭头状的
        // 依据当前拉动距离和最大拉动距离计算箭头大小值
        // 大小到一定值后开始折弯，计算箭头角度值
        val arrowZoom: Float = currentSlideLength / viewMaxLength // 箭头大小变化率
        val arrowAngle: Float = if (arrowZoom < 0.75f) 0f else (arrowZoom - 0.75f) * 2 // 箭头角度变化率
        // 箭头
        arrowPath.reset() // 先重置
        // 结合箭头大小值与箭头角度值设置折线路径
        arrowPath.moveTo(currentSlideLength / 2 + viewArrowSize * arrowAngle, viewHeight / 2 - arrowZoom * viewArrowSize)
        arrowPath.lineTo(currentSlideLength / 2 - viewArrowSize * arrowAngle, viewHeight / 2)
        arrowPath.lineTo(currentSlideLength / 2 + viewArrowSize * arrowAngle, viewHeight / 2 + arrowZoom * viewArrowSize)
        canvas.drawPath(arrowPath, arrowPaint)

        // 最多0.8透明度
        alpha = max(currentSlideLength / viewMaxLength - 0.2f, 0f)
    }

    /**
     * 更新当前拉动距离并重绘
     *
     * @param slideLength 当前拉动距离
     */
    fun updateSlideLength(slideLength: Float) {
        // 避免无意义重绘
        if (currentSlideLength == slideLength) return

        currentSlideLength = slideLength
        invalidate() // 会再次调用onDraw
    }

    /**
     * 给SlideBackIconView设置topMargin，起到定位效果
     *
     * @param position 触点位置
     */
    fun setSlideBackPosition(position: Float) {
        // 触点位置减去SlideBackIconView一半高度即为topMargin
        val layoutParams = FrameLayout.LayoutParams(layoutParams)
        layoutParams.topMargin = (position - viewHeight / 2).toInt()
        setLayoutParams(layoutParams)
    }
}